swifttool/manager.py

```diff
@@ -0,0 +1,288 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright (c) 2014, Blue Box Group, Inc.
+# Copyright (c) 2014, Craig Tracey <craigtracey@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+#
+
+import copy
+import logging
+import math
+import os
+import shutil
+import time
+import yaml
+
+from swifttool.ring_defintion import ringsdef_helper
+from swift.common.ring import RingBuilder
+from swifttool.utils import RING_TYPES, get_devices
+
+LOG = logging.getLogger(__name__)
+
+_STATUS_FILE = 'capacity.status'
+
+
+class CapacityManager(object):
+
+    def __init__(self, iterations=10,
+                 def_file='/etc/swift/ring_definition.yml',
+                 scaleup=True, metadata=None,
+                 outdir='/etc/swift/'):
+        self.iterations = int(iterations)
+        self.def_file = def_file
+        self.scaleup = scaleup
+        self.cluster_changes = {}
+        self.outdir = outdir
+        self.cluster_config = None
+        self.metadata = metadata
+        # Save the copy
+        shutil.copyfile(self.def_file, self.def_file + ".orig")
+        self.current_status = os.path.join(outdir, _STATUS_FILE)
+        self._write_status(0)
+        self._calculate_cluster_changes()
+
+    def __del__(self):
+        # Cleanup the files we create
+        os.remove(self.current_status)
+        os.remove(self.def_file + ".orig")
+
+    def update_cluster(self):
+        LOG.info("Starting to update the cluster")
+        min_part_hours = int(self._get_builder('account').min_part_hours)
+        for iteration in range(1, self.iterations + 1):
```

Comment:
`xrange` is the ideal choice for running this in Python 2 because it's a generator, rather than constructing the entire sequence.
If you're running this in py3k or you can guarantee that this will not be a ridiculously large number, then it's probably not an issue.